### ISSUE TITLE
Instability on WSS Connections, only works when one client abandon connection

### ISSUE THREAD

**Issue by Irahe** [TooTallNate/Java-WebSocket#466]
hello, im implementing a simple secure websocket server and client. My server implementation is the same as provided in the example folders, and im using websocket.org to test this implementation `(http://www.websocket.org/echo.html)`. I can connect using WSS. However, to do so, i need to request connection on thwo tabs, and close one of them. The other then, connects perfectly. If i dont close one of the tabs, all connections timeout. The server does not give me any exceptions on console not when it connects nor when it cant. 
I dont have many experience with WSS. However, WS implementaton works out of the box with this same server.
For my experience, seems like a endless loop is created when i try to connect. And only breaks when i kill one of the tabs trying to connect.

For the record, im using version `1.3.3`

Can you help me out pls?! =*

EDIT - 

Hello again, im posting a sample project to illustrate my problem. With this project is my keystore.jks.
The password for the keystore is `password` and the password for the key is `password`, as shown on my source.

Please, help out.
[d3_Servidor.zip](https://github.com/TooTallNate/Java-WebSocket/files/970624/d3_Servidor.zip)

**Comment by marci4** [comment]
Hello @Irahe,

I just tested your problem with my wss which I use in production and the issue mentioned does not occur for me.

Going to look into your sample project.

Til then, could you please tell me which clients you use? (so which browser)

Greetings
marci4

**Comment by marci4** [comment]
Hello @Irahe,

I looked into your project and for me it works fine.

The problem may be your self signed certificate.

Did you go to `https://localhost:8283` and told your browser to trust the certificate? (wss server has to run)

Greetings
marci4

**Comment by Irahe** [comment]
Hello @marci4 , thank you for your fast response!

Im using firefox 53.0 (32bits) for my tests. and i already have gone to https://localhost:8283 and accepted my certificate.
What i cant understand is that if the problem is with my certificate, why some times it can connect correctly?

The client im testig with is from websocket.org. it is on this link: `http://www.websocket.org/echo.html`.

Can you tell me step by step how i must configure my setup so it works for me too?

Thank you =*

**Comment by marci4** [comment]
Hello @Irahe,

this is interesting. This problem occurs for me as well in firefox, but not in chrome.
It is also limited to localhost, since I was able to connect to my production wss.

Just to clear some things up, your client is firefox.
The  websocket.org website just provides an easy interface for the javascript implementation below.

Gonna test it against other implementations of websocket servers and gonna keep you updated.

Greetings
marci4

**Comment by Irahe** [comment]
Hello there, @marci4 , 

Thank you again for your immediate response! This is really getting interesting.
Im getting a Let's Encrypt certificate for my production domain, and so ill try to apply it to a test server on a subdomain there... 
I hope it works...

I keep you posted!

Best wishes,
Irahe Kasprzykowski.

**Comment by Irahe** [comment]
Hello @marci4,

I have acquired a certificate from Let's Encrypt, and have tested this same issue with the chatserver on my domain directly.
Unfortunately it still does not work properly. I have tested it on Edge, IE and Firefox, and none of these works.
My server is now running on wss://www.douli3.com.br:8283
If you can, please test it out.

Im sending you my implementation for SSLContext of Lets encrypt certificate, with the chatserver, however, i cant send you the 'pem' files. wich are on the root server...
I know they are been accessed correctly because some times im able to connect with the server...

[d3_ServidorLE.zip](https://github.com/TooTallNate/Java-WebSocket/files/970911/d3_ServidorLE.zip)



Greetings,

Irahe Kasprzykowski

**Comment by marci4** [comment]
Ok here is my current status:

SSC = Self signed certificate
LE = Let's Encrypt

Test case echo.websocket.org SSC:
- Internet Explorer 11.1066 **ERROR** Security error
- Firefox 53.0 **ERROR** Sometimes doesn't work (like 75% are fine)
- Edge 38.14393 **ERROR** Security error
- Chrome 59.0.3071.29 **PASS** 

Test case local file SSC:
- Internet Explorer 11.1066 **PASS** ALWAYS
- Firefox 53.0 **ERROR** Sometimes doesn't work (like 75% are fine)
- Edge 38.14393 **PASS** Cannot get the browser to trust the certificate
- Chrome 59.0.3071.29 **PASS**

Test case local file LE:
- Internet Explorer 11.1066 **PASS** (After not checking for revoke server)
- Firefox 53.0 ERROR Sometimes doesn't work (like 95% are fine)
- Edge 38.14393 **PASS** (After not checking for revoke server)
- Chrome 59.0.3071.29 **PASS**

I also was able to connect to your server without problems.

Greetings
marci4

**Comment by Irahe** [comment]
Hello @marci4 ,

The server on my domain is working on debug mode, so, ill be posting the log of your tests here so you can see what happened server side.

I also can connect to the server via chrome (seems to let absolitely every single thing pass thru). However, when i try to connect via Firefox, IE or Edge it times out.

I figgered a way to trigger the problem, but not on chrome.

If you connect via websocket.org, and immediately disconnect and try to immediately reconnect, it wont connect to the server.

Its realy strange that this is a browser support problem since they only follow the protocon, and my certificate on www.douli3.com.br is up to date and active.

Do you have some workaround that i can do to make it work? Any tips?


Best wishes,

Irahe Kasprzykowski

**Comment by marci4** [comment]
Hello @Irahe,

as mentioned above, do you really need tcpNoDelay set?
It does cause additional stress for the cpu and network, and I think it could be the main reason for the problems.
It is the only difference between your and my server.

Greetings
marci4

**Comment by Irahe** [comment]
Hi, @marci4 ,

I'll disabled TCPNoDelay on the chatserver at my domain. 
However, my problems with Firefox and other naughty browsers other than chrome keep happening.
In Firefox, if i connect, disconnect and try to reconnect, it just does nothing for a while, then times out.

Why is Firefox so unrighteous?  LOL...


Greetings,

Irahe Kasprzykowski

**Comment by zhoulifu** [comment]
Hey guys, I've test this issue using websocket.org with a [netty websocket server](https://gist.github.com/zhoulifu/788287805d7ec4de8c53844db98beb39), it works fine with firefox as the same as chrome. I believe this problem has nothing to do with the self-signed certificate.

Since it's sometimes failed to connect to server due to timeout, is it possible that this problem was caused by some resources leak?

**Comment by Irahe** [comment]
@zhoulifu , thank you for the advice.
I'll take a look at netty lib. However a already have made a cross-solution, proxying wss:// to ws:// in apache server, this way i can still have a https connection and a ws one.

=D

**Comment by zhoulifu** [comment]
@Irahe Glad to hear you've solved your problem.
But it will have performance effect I think because apache needs to hold extra socket for every ws connection. Just mentioning :)

**Comment by marci4** [comment]
Could be related to #385 and #160

**Comment by marci4** [comment]
It actually is related to #385.

Check out http://stackoverflow.com/a/30760134 for a more detail explanation.
TL;DR java doesn't support `TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256` and firefox requests this.

I quess I gonna patch this directly.

Greetings
marci4

**Comment by Irahe** [comment]
@marci4 , FANTASTIC!
The 1.3.4 is already available? I can just download the source and rebuild?

Thank you guys again!

Irahe Kasprzykowski

**Comment by marci4** [comment]
Hello @Irahe,

you can build the jar directly from the sources.
1.3.4 will not be around for at least the next 2 weeks probably a bit more. (Don't wanna go be at version 1.33.7 at the end of the year ;) )

Greetings
marci4

**Comment by Irahe** [comment]
@marci4 ,

The changes needed are already at the source in here( `https://github.com/TooTallNate/Java-WebSocket`) or i should go to your personal fork at (`https://github.com/marci4/Java-WebSocket-Dev`)?

Thanks,

I hope the 2.0 reaches us soon! =D

**Comment by marci4** [comment]
Hello @Irahe,

no I already merged them into this repository.

hahaha probably not ;)

**Comment by Irahe** [comment]
@marci4  and @zhoulifu ,

Thank you for your help!
I just imported the new lib into my mais project and it worked like a champ!
You can tag this issue as FIXED LIKE A BOSS...  XD

`tausend dank!`,
Irahe Kasprzykowski

**Comment by marci4** [comment]
Happy to hear that! 

Closing issue.
Greetings
marci4
